hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientJCacheTypeChecks.java

```diff
@@ -0,0 +1,373 @@
+        join.getMulticastConfig().setEnabled(false);
+        join.getTcpIpConfig().setEnabled(true);
+        join.getTcpIpConfig().addMember("127.0.0.1");
+
+        hz1 = Hazelcast.newHazelcastInstance(config);
+        hz2 = Hazelcast.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getNetworkConfig().addAddress("127.0.0.1");
+        client = HazelcastClient.newHazelcastClient(clientConfig);
+
+        cachingProvider1 = HazelcastServerCachingProvider.createCachingProvider(hz1);
+        cachingProvider2 = HazelcastServerCachingProvider.createCachingProvider(hz2);
+        cachingProvider3 = HazelcastClientCachingProvider.createCachingProvider(client);
+    }
+
+    @AfterClass
+    public static void tear() {
+        cachingProvider1.close();
+        cachingProvider2.close();
+        cachingProvider3.close();
+        HazelcastClient.shutdownAll();
+        Hazelcast.shutdownAll();
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_put() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider3.getCacheManager();
+        assertNotNull(cacheManager);
+
+        assertNull(cacheManager.getCache(cacheName));
+
+        CompleteConfiguration<Integer, String> config = new CacheConfig<Integer, String>()
+                .setTypes(Integer.class, String.class);
+
+        Cache<Integer, String> cache = cacheManager.createCache(cacheName, config);
+        assertNotNull(cache);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                CacheManager cm1 = cachingProvider1.getCacheManager();
+                assertNotNull(cm1.getCache(cacheName, Integer.class, String.class));
+                CacheManager cm2 = cachingProvider2.getCacheManager();
+                assertNotNull(cm2.getCache(cacheName, Integer.class, String.class));
+            }
+        });
+
+        Cache rawCache = (Cache) cache;
+        rawCache.put(new Object(), new Object());
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_get() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider3.getCacheManager();
+        assertNotNull(cacheManager);
+
+        assertNull(cacheManager.getCache(cacheName));
+
+        CompleteConfiguration<Integer, String> config = new CacheConfig<Integer, String>();
+
+        Cache<Integer, String> cache = cacheManager.createCache(cacheName, config);
+        assertNotNull(cache);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                CacheManager cm1 = cachingProvider1.getCacheManager();
+                assertNotNull(cm1.getCache(cacheName));
```

Comment:
there are many duplications, especially almost every test has this part.
is it possible make shorter it? like making inner class and creating new one every time. or some another solution.